File: unity/Assets/Scripts/Game.cs
using System;
using System.Collections.Generic;

namespace Minesweeper.Unity
{
    public enum CellState { Hidden, Revealed, Flagged, Questioned }
    public enum GameStatus { Idle, Playing, Won, Lost }

    public class Cell
    {
        public int Row, Col, Adjacent;
        public bool Mine, Exploded, WrongFlag;
        public CellState State = CellState.Hidden;
        public Cell(int r, int c) { Row = r; Col = c; }
    }

    public struct Preset
    {
        public string Name;
        public int Rows, Cols, Mines;
        public Preset(string n, int r, int c, int m) { Name = n; Rows = r; Cols = c; Mines = m; }
        public static readonly Preset[] All = {
            new Preset("Beginner", 9, 9, 10),
            new Preset("Intermediate", 16, 16, 40),
            new Preset("Expert", 16, 30, 99),
        };
    }

    public class Game
    {
        public int Rows, Cols, Mines;
        public Cell[,] Cells;
        public int Flags;
        public bool FirstClick = true;
        public GameStatus Status = GameStatus.Idle;
        public DateTime? StartedAt, EndedAt;
        private readonly System.Random _rand = new System.Random();

        public Game(Preset p)
        {
            Rows = p.Rows; Cols = p.Cols; Mines = p.Mines;
            Cells = new Cell[Rows, Cols];
            for (int r = 0; r < Rows; r++)
                for (int c = 0; c < Cols; c++) Cells[r, c] = new Cell(r, c);
        }

        public int MinesRemaining { get { return Math.Max(Mines - Flags, 0); } }

        public int ElapsedSeconds
        {
            get
            {
                if (!StartedAt.HasValue) return 0;
                var end = (Status == GameStatus.Won || Status == GameStatus.Lost) && EndedAt.HasValue
                    ? EndedAt.Value : DateTime.UtcNow;
                return Math.Min((int)(end - StartedAt.Value).TotalSeconds, 999);
            }
        }

        bool InB(int r, int c) { return r >= 0 && r < Rows && c >= 0 && c < Cols; }

        void PlaceMines(int sr, int sc)
        {
            var safe = new HashSet<int>();
            for (int dr = -1; dr <= 1; dr++)
                for (int dc = -1; dc <= 1; dc++)
                    if (InB(sr + dr, sc + dc)) safe.Add((sr + dr) * Cols + (sc + dc));
            var cands = new List<int>();
            for (int r = 0; r < Rows; r++)
                for (int c = 0; c < Cols; c++)
                    if (!safe.Contains(r * Cols + c)) cands.Add(r * Cols + c);
            for (int i = cands.Count - 1; i > 0; i--)
            {
                int j = _rand.Next(i + 1);
                var t = cands[i]; cands[i] = cands[j]; cands[j] = t;
            }
            int n = Math.Min(Mines, cands.Count);
            for (int i = 0; i < n; i++)
            {
                int idx = cands[i];
                Cells[idx / Cols, idx % Cols].Mine = true;
            }
            for (int r = 0; r < Rows; r++)
                for (int c = 0; c < Cols; c++)
                {
                    if (Cells[r, c].Mine) continue;
                    int k = 0;
                    for (int dr = -1; dr <= 1; dr++)
                        for (int dc = -1; dc <= 1; dc++)
                            if ((dr | dc) != 0 && InB(r + dr, c + dc) && Cells[r + dr, c + dc].Mine) k++;
                    Cells[r, c].Adjacent = k;
                }
        }

        public void Reveal(int r, int c)
        {
            if (Status == GameStatus.Won || Status == GameStatus.Lost) return;
            var cell = Cells[r, c];
            if (cell.State == CellState.Revealed || cell.State == CellState.Flagged) return;
            if (FirstClick) { PlaceMines(r, c); FirstClick = false; Status = GameStatus.Playing; StartedAt = DateTime.UtcNow; }
            if (cell.Mine)
            {
                cell.State = CellState.Revealed; cell.Exploded = true;
                RevealAll(); Status = GameStatus.Lost; EndedAt = DateTime.UtcNow; return;
            }
            Flood(r, c);
            if (CheckWin()) { Status = GameStatus.Won; EndedAt = DateTime.UtcNow; }
        }

        void Flood(int r, int c)
        {
            var stack = new Stack<int>();
            stack.Push(r * Cols + c);
            while (stack.Count > 0)
            {
                int p = stack.Pop();
                int rr = p / Cols, cc = p % Cols;
                var cell = Cells[rr, cc];
                if (cell.State == CellState.Revealed || cell.State == CellState.Flagged || cell.Mine) continue;
                cell.State = CellState.Revealed;
                if (cell.Adjacent > 0) continue;
                for (int dr = -1; dr <= 1; dr++)
                    for (int dc = -1; dc <= 1; dc++)
                        if ((dr | dc) != 0 && InB(rr + dr, cc + dc))
                            stack.Push((rr + dr) * Cols + (cc + dc));
            }
        }

        public void ToggleFlag(int r, int c)
        {
            if (Status == GameStatus.Won || Status == GameStatus.Lost) return;
            var cell = Cells[r, c];
            if (cell.State == CellState.Revealed) return;
            if (cell.State == CellState.Hidden) { cell.State = CellState.Flagged; Flags++; }
            else if (cell.State == CellState.Flagged) { cell.State = CellState.Questioned; Flags--; }
            else cell.State = CellState.Hidden;
        }

        public void Chord(int r, int c)
        {
            if (Status != GameStatus.Playing) return;
            var cell = Cells[r, c];
            if (cell.State != CellState.Revealed || cell.Adjacent == 0) return;
            int flagged = 0;
            var hidden = new List<(int, int)>();
            for (int dr = -1; dr <= 1; dr++)
                for (int dc = -1; dc <= 1; dc++)
                {
                    if ((dr | dc) == 0) continue;
                    int nr = r + dr, nc = c + dc;
                    if (!InB(nr, nc)) continue;
                    var n = Cells[nr, nc];
                    if (n.State == CellState.Flagged) flagged++;
                    else if (n.State != CellState.Revealed) hidden.Add((nr, nc));
                }
            if (flagged != cell.Adjacent) return;
            foreach (var h in hidden) { Reveal(h.Item1, h.Item2); if (Status == GameStatus.Lost) return; }
        }

        void RevealAll()
        {
            for (int r = 0; r < Rows; r++)
                for (int c = 0; c < Cols; c++)
                {
                    var cell = Cells[r, c];
                    if (cell.Mine && cell.State != CellState.Flagged) cell.State = CellState.Revealed;
                    if (!cell.Mine && cell.State == CellState.Flagged) { cell.WrongFlag = true; cell.State = CellState.Revealed; }
                }
        }

        bool CheckWin()
        {
            for (int r = 0; r < Rows; r++)
                for (int c = 0; c < Cols; c++)
                    if (!Cells[r, c].Mine && Cells[r, c].State != CellState.Revealed) return false;
            for (int r = 0; r < Rows; r++)
                for (int c = 0; c < Cols; c++)
                    if (Cells[r, c].Mine && Cells[r, c].State != CellState.Flagged)
                    { Cells[r, c].State = CellState.Flagged; Flags++; }
            return true;
        }
    }
}

File: ios/Sources/MinesweeperCore/Game.swift
import Foundation

public enum CellState { case hidden, revealed, flagged, questioned }
public enum GameStatus { case idle, playing, won, lost }

public final class Cell {
    public let row: Int
    public let col: Int
    public var mine = false
    public var state: CellState = .hidden
    public var adjacent = 0
    public var exploded = false
    public var wrongFlag = false
    public init(_ r: Int, _ c: Int) { row = r; col = c }
}

public struct Preset: Hashable {
    public let name: String
    public let rows: Int
    public let cols: Int
    public let mines: Int
    public init(_ n: String, _ r: Int, _ c: Int, _ m: Int) { name = n; rows = r; cols = c; mines = m }
    public static let beginner = Preset("Beginner", 9, 9, 10)
    public static let intermediate = Preset("Intermediate", 16, 16, 40)
    public static let expert = Preset("Expert", 16, 30, 99)
    public static let all: [Preset] = [.beginner, .intermediate, .expert]
}

public final class Game {
    public let rows: Int
    public let cols: Int
    public let mines: Int
    public private(set) var cells: [[Cell]]
    public private(set) var flags = 0
    public private(set) var firstClick = true
    public private(set) var status: GameStatus = .idle
    public private(set) var startedAt: Date?
    public private(set) var endedAt: Date?

    public init(_ p: Preset) {
        rows = p.rows; cols = p.cols; mines = p.mines
        cells = (0..<p.rows).map { r in (0..<p.cols).map { c in Cell(r, c) } }
    }

    public var minesRemaining: Int { max(mines - flags, 0) }
    public var elapsedSeconds: Int {
        guard let s = startedAt else { return 0 }
        let end: Date
        if status == .won || status == .lost, let e = endedAt { end = e } else { end = Date() }
        return min(Int(end.timeIntervalSince(s)), 999)
    }

    private func inB(_ r: Int, _ c: Int) -> Bool { r >= 0 && r < rows && c >= 0 && c < cols }

    private func placeMines(safeRow sr: Int, safeCol sc: Int) {
        var safe = Set<Int>()
        for dr in -1...1 { for dc in -1...1 where inB(sr+dr, sc+dc) { safe.insert((sr+dr)*cols + (sc+dc)) } }
        var cands: [Int] = []
        for r in 0..<rows { for c in 0..<cols where !safe.contains(r*cols + c) { cands.append(r*cols + c) } }
        cands.shuffle()
        for idx in cands.prefix(min(mines, cands.count)) {
            cells[idx / cols][idx % cols].mine = true
        }
        for r in 0..<rows {
            for c in 0..<cols where !cells[r][c].mine {
                var n = 0
                for dr in -1...1 { for dc in -1...1 where (dr|dc) != 0 && inB(r+dr, c+dc) && cells[r+dr][c+dc].mine { n += 1 } }
                cells[r][c].adjacent = n
            }
        }
    }

    public func reveal(_ r: Int, _ c: Int) {
        guard status != .won && status != .lost else { return }
        let cell = cells[r][c]
        guard cell.state != .revealed && cell.state != .flagged else { return }
        if firstClick {
            placeMines(safeRow: r, safeCol: c)
            firstClick = false
            status = .playing
            startedAt = Date()
        }
        if cell.mine {
            cell.state = .revealed; cell.exploded = true
            revealAll(); status = .lost; endedAt = Date(); return
        }
        flood(r, c)
        if checkWin() { status = .won; endedAt = Date() }
    }

    private func flood(_ r: Int, _ c: Int) {
        var stack = [(r, c)]
        while let (rr, cc) = stack.popLast() {
            let cell = cells[rr][cc]
            if cell.state == .revealed || cell.state == .flagged || cell.mine { continue }
            cell.state = .revealed
            if cell.adjacent > 0 { continue }
            for dr in -1...1 { for dc in -1...1 where (dr|dc) != 0 && inB(rr+dr, cc+dc) { stack.append((rr+dr, cc+dc)) } }
        }
    }

    public func toggleFlag(_ r: Int, _ c: Int) {
        guard status != .won && status != .lost else { return }
        let cell = cells[r][c]
        guard cell.state != .revealed else { return }
        switch cell.state {
        case .hidden: cell.state = .flagged; flags += 1
        case .flagged: cell.state = .questioned; flags -= 1
        case .questioned: cell.state = .hidden
        case .revealed: break
        }
    }

    public func chord(_ r: Int, _ c: Int) {
        guard status == .playing else { return }
        let cell = cells[r][c]
        guard cell.state == .revealed && cell.adjacent > 0 else { return }
        var flagged = 0
        var hidden: [(Int, Int)] = []
        for dr in -1...1 {
            for dc in -1...1 where (dr|dc) != 0 {
                let nr = r+dr, nc = c+dc
                guard inB(nr, nc) else { continue }
                let n = cells[nr][nc]
                if n.state == .flagged { flagged += 1 }
                else if n.state != .revealed { hidden.append((nr, nc)) }
            }
        }
        guard flagged == cell.adjacent else { return }
        for (hr, hc) in hidden { reveal(hr, hc); if status == .lost { return } }
    }

    private func revealAll() {
        for row in cells {
            for c in row {
                if c.mine && c.state != .flagged { c.state = .revealed }
                if !c.mine && c.state == .flagged { c.wrongFlag = true; c.state = .revealed }
            }
        }
    }

    private func checkWin() -> Bool {
        for row in cells { for c in row where !c.mine && c.state != .revealed { return false } }
        for row in cells {
            for c in row where c.mine && c.state != .flagged { c.state = .flagged; flags += 1 }
        }
        return true
    }
}
